src/Storages/ReadInOrderOptimizer.cpp

```diff
@@ -180,46 +198,54 @@ InputOrderInfoPtr ReadInOrderOptimizer::getInputOrderImpl(
     const StorageMetadataPtr & metadata_snapshot,
     const SortDescription & description,
     const ManyExpressionActions & actions,
+    const ContextPtr & context,
     UInt64 limit) const
 {
     auto sorting_key_columns = metadata_snapshot->getSortingKeyColumns();
     int read_direction = description.at(0).direction;
 
-    size_t fixed_prefix_size = calculateFixedPrefixSize(query, sorting_key_columns);
-    size_t descr_prefix_size = std::min(description.size(), sorting_key_columns.size() - fixed_prefix_size);
+    auto fixed_sorting_columns = getFixedSortingColumns(query, sorting_key_columns, context);
+
+    SortDescription sort_description_for_merging;
+    sort_description_for_merging.reserve(description.size());
 
-    SortDescription order_key_prefix_descr;
-    order_key_prefix_descr.reserve(descr_prefix_size);
+    size_t desc_pos = 0;
+    size_t key_pos = 0;
 
-    for (size_t i = 0; i < descr_prefix_size; ++i)
+    while (desc_pos < description.size() && key_pos < sorting_key_columns.size())
     {
-        if (forbidden_columns.contains(description[i].column_name))
+        if (forbidden_columns.contains(description[desc_pos].column_name))
             break;
 
-        int current_direction = matchSortDescriptionAndKey(
-            actions[i]->getActions(), description[i], sorting_key_columns[i + fixed_prefix_size]);
+        int current_direction = matchSortDescriptionAndKey(actions[desc_pos]->getActions(), description[desc_pos], sorting_key_columns[key_pos]);
+        bool is_matched = current_direction && (desc_pos == 0 || current_direction == read_direction);
+
+        if (!is_matched)
+        {
+            /// If one of the sorting columns is constant after filtering,
+            /// skip it, because it won't affect order anymore.
+            if (fixed_sorting_columns.contains(sorting_key_columns[key_pos]))
```

Comment:
It seems like we can skip it even if `is_matched` is true, can't we?

Reply:
If we skip when `is_matched` is true, `sort_description_for_merging` will probably differ from prefix of global sort description and we will get mismatch of sort descriptions in `FinishSorting` transform. We need somehow to modify global sort description and it's hard to do from this place. I think, it's ok to leave it as is.